autorest/responder_test.go

```diff
@@ -423,6 +424,60 @@ func TestRespondAcceptsNullResponse(t *testing.T) {
 	}
 }
 
+func TestWithErrorUnlessStatusCodeOKResponse(t *testing.T) {
+	v := &mocks.T{}
+	r := mocks.NewResponseWithContent(jsonT)
+	err := Respond(r,
+		WithErrorUnlessStatusCode(http.StatusOK),
+		ByUnmarshallingJSON(v),
+		ByClosing())
+
+	if err != nil {
+		t.Fatalf("autorest: WithErrorUnlessStatusCode(http.StatusOK) failed on okay response. (%v)", err)
+	}
+
+	if v.Name != "Rob Pike" || v.Age != 42 {
+		t.Fatalf("autorest: WithErrorUnlessStatusCode(http.StatusOK) corrupted the response body of okay response.")
+	}
+}
+
+func TesWithErrorUnlessStatusCodeErrorResponse(t *testing.T) {
+	v := &mocks.T{}
+	e := &mocks.T{}
+	r := mocks.NewResponseWithContent(jsonT)
+	r.Status = "400 BadRequest"
+	r.StatusCode = http.StatusBadRequest
+
+	err := Respond(r,
+		WithErrorUnlessStatusCode(http.StatusOK),
+		ByUnmarshallingJSON(v),
+		ByClosing())
+
+	if err == nil {
+		t.Fatalf("autorest: WithErrorUnlessStatusCode(http.StatusOK) failed to return error on non-okay response.")
+	}
+
+	var errorRespBody []byte
+	if derr, ok := err.(DetailedError); !ok {
+		t.Fatalf("autorest: WithErrorUnlessStatusCode(http.StatusOK) failed to return detailed error on non-okay response. (%v)", err)
+	} else {
+		errorRespBody = derr.ServiceError
+	}
+
+	if errorRespBody == nil {
+		t.Fatalf("autorest: WithErrorUnlessStatusCode(http.StatusOK) failed to set ServiceError of detailed error on non-okay response.")
+	}
+
+	err = json.Unmarshal(errorRespBody, e)
+	if err == nil {
+		t.Fatalf("autorest: WithErrorUnlessStatusCode(http.StatusOK) corrupted the response body on non-okay response. (%v)", err)
+	}
+
+	if e.Name != "Rob Pike" || e.Age != 42 {
+		t.Fatalf("autorest: WithErrorUnlessStatusCode(http.StatusOK) corrupted the response body on non-okay response.")
```

Comment:
you are failing  with the same error here and above. just change it to:

```
if expected := T{Name:"Rob Pike", Age: 42}; e != expected {
    t.Fatalf("wrong value from parsed ServiceError: got=%#v expected=%#v", e, expected)
}
```

This will make it clear what's going on.